### ISSUE TITLE
Next paper candidates: 14 June

### ISSUE THREAD

**Issue by bzz** [src-d/reading-club#61]
# Next paper candidates
Let's propose papers to study next! All papers mentioned in the comments of this issue will be listed in the next vote.

## Last session runner-up(s)
- [Graph Attention Networks](https://arxiv.org/abs/1710.10903)
    > We present graph attention networks (GATs), novel neural network architectures that operate on graph-structured data, leveraging masked self-attentional layers to address the shortcomings of prior methods based on graph convolutions or their approximations. By stacking layers in which nodes are able to attend over their neighborhoods' features, we enable (implicitly) specifying different weights to different nodes in a neighborhood, without requiring any kind of costly matrix operation (such as inversion) or depending on knowing the graph structure upfront. In this way, we address several key challenges of spectral-based graph neural networks simultaneously, and make our model readily applicable to inductive as well as transductive problems. Our GAT models have achieved or matched state-of-the-art results across four established transductive and inductive graph benchmarks: the Cora, Citeseer and Pubmed citation network datasets, as well as a protein-protein interaction dataset (wherein test graphs remain unseen during training).

- [Graph Matching Networks for Learning the Similarity of Graph Structured Objects](https://arxiv.org/abs/1904.12787)
    > This paper addresses the challenging problem of retrieval and matching of graph structured objects, and makes two key contributions. First, we demonstrate how Graph Neural Networks (GNN), which have emerged as an effective model for various supervised prediction problems defined on structured data, can be trained to produce embedding of graphs in vector spaces that enables efficient similarity reasoning. Second, we propose a novel Graph Matching Network model that, given a pair of graphs as input, computes a similarity score between them by jointly reasoning on the pair through a new cross-graph attention-based matching mechanism. We demonstrate the effectiveness of our models on different domains including the challenging problem of control-flow-graph based function similarity search that plays an important role in the detection of vulnerabilities in software systems. The experimental analysis demonstrates that our models are not only able to exploit structure in the context of similarity learning but they can also outperform domain-specific baseline systems that have been carefully hand-engineered for these problems.

**Comment by kuba--** [comment]
I propose this one "Coloring Big Graphs with AlphaGoZero" - (thanks @m09): https://arxiv.org/abs/1902.10162
> We show that recent innovations in deep reinforcement learning can effectively color very
large graphs – a well-known NP-hard problem
with clear commercial applications. Because the
Monte Carlo Tree Search with Upper Confidence
Bound algorithm used in AlphaGoZero can improve the performance of a given heuristic, our approach allows deep neural networks trained using
high performance computing (HPC) technologies
to transform computation into improved heuristics
with zero prior knowledge. Key to our approach
is the introduction of a novel deep neural network
architecture (FastColorNet) that has access to the
full graph context and requires O(V ) time and
space to color a graph with V vertices, which enables scaling to very large graphs that arise in real
applications like parallel computing, compilers,
numerical solvers, and design automation, among
others. As a result, we are able to learn new state
of the art heuristics for graph coloring.

**Comment by m09** [comment]
[Import2vec: learning embeddings for software libraries](https://arxiv.org/abs/1904.03990)

> We consider the problem of developing suitable learning representations (embeddings) for library packages that capture semantic similarity among libraries. Such representations are known to improve the performance of downstream learning tasks (e.g. classification) or applications such as contextual search and analogical reasoning.
> 
> We apply word embedding techniques from natural language processing (NLP) to train embeddings for library packages (“library vectors”). Library vectors represent libraries by similar context of use as determined by import statements present in source code. Experimental results obtained from training such embeddings on three large open source software corpora reveals that library vectors capture semantically meaningful relationships among software libraries, such as the relationship between frameworks and their plug-ins and libraries commonly used together within ecosystems such as big data infrastructure projects (in Java), front-end and back-end web development frameworks (in JavaScript) and data science toolkits (in Python).
> 
> For a live demo of a contextual search engine built on import2vec, please visit https://www.code-compass.com.

**Comment by m09** [comment]
[Revisiting Graph Neural Networks: All We Have is Low-Pass Filters](https://arxiv.org/abs/1905.09550)

> Graph neural networks have become one of the most important techniques to solve machine learning problems on graph-structured data. Recent work on vertex classification proposed deep and distributed learning models to achieve high performance and scalability. However, we find that the feature vectors of benchmark datasets are already quite informative for the classification task, and the graph structure only provides a means to denoise the data. In this paper, we develop a theoretical framework based on graph signal processing for analyzing graph neural networks. Our results indicate that graph neural networks only perform low-pass filtering on feature vectors and do not have the non-linear manifold learning property. We further investigate their resilience to feature noise and propose some insights on GCN-based graph neural network design.

**Comment by xennygrimmato** [comment]
[Towards Neural Decompilation](https://arxiv.org/abs/1905.08325)
> We address the problem of automatic decompilation, converting a program in low-level representation back to a higher-level human-readable programming language. The problem of decompilation is extremely important for security researchers. Finding vulnerabilities and understanding how malware operates is much easier when done over source code. 
The importance of decompilation has motivated the construction of hand-crafted rule-based decompilers. Such decompilers have been designed by experts to detect specific control-flow structures and idioms in low-level code and lift them to source level. The cost of supporting additional languages or new language features in these models is very high. 
We present a novel approach to decompilation based on neural machine translation. The main idea is to automatically learn a decompiler from a given compiler. Given a compiler from a source language S to a target language T , our approach automatically trains a decompiler that can translate (decompile) T back to S . We used our framework to decompile both LLVM IR and x86 assembly to C code with high success rates. Using our LLVM and x86 instantiations, we were able to successfully decompile over 97% and 88% of our benchmarks respectively.

**Comment by kuba--** [comment]
"Evolution Strategies as a Scalable Alternative to Reinforcement Learning": https://arxiv.org/pdf/1703.03864.pdf
> We explore the use of Evolution Strategies (ES), a class of black box optimization algorithms, as an alternative to popular MDP-based RL techniques such as Q- learning and Policy Gradients. Experiments on MuJoCo and Atari show that ES is a viable solution strategy that scales extremely well with the number of CPUs available: By using a novel communication strategy based on common random numbers, our ES implementation only needs to communicate scalars, making it possible to scale to over a thousand parallel workers. This allows us to solve 3D humanoid walking in 10 minutes and obtain competitive results on most Atari games after one hour of training. In addition, we highlight several advantages of ES as a black box optimization technique: it is invariant to action frequency and delayed rewards, tolerant of extremely long horizons, and does not need temporal discounting or value function approximation.

**Comment by kuba--** [comment]
"Yet Another Efficient Unification Algorithm": https://arxiv.org/pdf/cs/0603080.pdf
> The unification algorithm is at the core of the logic programming paradigm, the first unification algorithm being developed by Robinson [5]. More efficient algorithms were developed later [4] and I introduce here yet another efficient unification algorithm centered on a specific data structure, called the Unification Table.

**Comment by bzz** [comment]
Posting the vote on Slack a bit later today.

As there were no votes [last time](https://sourced-community.slack.com/archives/CCLPL1WQH/p1558695052001300) for any of the papers, I'll consider the runner-up section to be empty and is not going to conclude it into vote this time.

**Comment by m09** [comment]
[Posted](https://sourced-community.slack.com/archives/CCLPL1WQH/p1559923847001400) on Slack.

**Comment by bzz** [comment]
`Coloring Big Graphs with AlphaGoZero` was chosen!
